Jupyter Notebook to Blog Post GitHub Action

hamelsmu opened the pull request:
First of all, thank you to @sgugger for putting up with several questions and making amendments to the nbdev library to make this possible.  

This GitHub Action uses nbdev to convert notebooks to HTML, and saves these as .md files in `_posts` so that they can be rendered by GitHub pages.  The motivation is that folks can write their blog posts as notebooks and simply save these notebooks into a folder and it magically becomes a blog post.  

### High-Level Overview
1.  Save a Jupyter Notebook into the `_notebooks` folder, preferably prepended with `YYYY-MM-DD` as you would do with a markdown post.  
2.  The authors should put the following front matter in the notebook per usual 
```
# Title
> Summary
- metadata_key1: metadata_value1
- metadata_key2: metadata_value2 
```

3.  **If you fail to use the right naming convention**, GitHub Actions will automatically prepend `YYYY-MM-DD` to the `.md` file placed in `_posts`, which is inferred by the file's last modified date.  Even though this file has a `.md` extension, it is really a HTML file with front matter (depending on your point of view).
4.  This workflow triggers only when a change to a file in `_notebooks` is detected on the master branch.
5. **UPDATE**:  It could be annoying to have Actions automatically do a notebook conversion on every push to master since that would cause your local environment to have conflicts and you would have to pull after every push when editing notebooks.  In order to circumvent this problem, **this workflow only triggers when the magic word `/sync` appears anywhere in your commit messages** ( see [filter](https://github.com/fastai/fast_template/pull/5/files#diff-fd48c7c54988c6b2d00efd3eaf20d1abR11) ).  Also, as mentioned above this workflow will only trigger when you push to the master branch.  We can change and customize any of these behaviors as you wish.
6. Since the conversion is done as HTML, interactive charts and graphs in Altair are preserved.  (Still trying to figure out ipywidgets)


### Other Notes / Discussion:
- Right now if the destination file is prepended with `YYYY-MM-DD` automatically because the notebook is not named correctly, the only feedback that the user gets are the logs in Actions.  Perhaps I should have the Action open a GitHub issue informing the user that they haven't followed convention and the date was inferred.  ( Note that the notebook's names are never changed, only the converted file).
- when an error occurs with notebook2html, it only prints the error to the screen, however it might be nice to raise the error so that the workflow fails and an alert is thrown via the normal mechanisms? I made a PR to address in case it is helpful https://github.com/fastai/nbdev/pull/65

hamelsmu committed:
turn on actions

hamelsmu committed:
turn on actions

hamelsmu committed:
fix workflow

hamelsmu committed:
fix workflow

Commit:
Update Jupyter Blog Posts

hamelsmu committed:
Merge pull request #1 from hamelsmu/nb2postAction

Nb2post action

hamelsmu committed:
troubleshoot

hamelsmu committed:
Merge branch 'nb2postAction' of https://github.com/hamelsmu/fast_template into nb2postAction

Commit:
Update Jupyter Blog Posts

hamelsmu committed:
Merge pull request #2 from hamelsmu/nb2postAction

Nb2post action

Commit:
Update Jupyter Blog Posts

hamelsmu committed:
correct path error

hamelsmu committed:
Merge branch 'master' of https://github.com/hamelsmu/fast_template

Commit:
Update Jupyter Blog Posts

hamelsmu committed:
troubleshoot

hamelsmu committed:
Merge branch 'master' of https://github.com/hamelsmu/fast_template

hamelsmu committed:
troubleshoot

hamelsmu committed:
troubleshoot

hamelsmu committed:
troubleshoot

hamelsmu committed:
troubleshoot

hamelsmu committed:
troubleshoot

hamelsmu committed:
troubleshoot

hamelsmu committed:
troubleshoot

hamelsmu committed:
troubleshoot

hamelsmu committed:
troubleshoot

Commit:
Update Jupyter Blog Posts

hamelsmu committed:
troubleshoot

hamelsmu committed:
Merge branch 'master' of https://github.com/hamelsmu/fast_template

Commit:
Update Jupyter Blog Posts

hamelsmu committed:
troubleshoot

hamelsmu committed:
Merge branch 'master' of https://github.com/hamelsmu/fast_template

hamelsmu committed:
troubleshoot

Commit:
Update Jupyter Blog Posts

hamelsmu committed:
save

hamelsmu committed:
Update nb2post.py

Commit:
Update Jupyter Blog Posts

hamelsmu committed:
made a change to nb

hamelsmu committed:
erge branch 'master' of https://github.com/hamelsmu/fast_template

Commit:
Update Jupyter Blog Posts

hamelsmu committed:
add filters to workflow

hamelsmu committed:
Merge branch 'master' of https://github.com/hamelsmu/fast_template

hamelsmu committed:
made a change to nb

hamelsmu committed:
made a change to nb

hamelsmu committed:
testing renaming

Commit:
Update Jupyter Blog Posts

hamelsmu committed:
remove renaming test artifacts

Commit:
Update Jupyter Blog Posts

hamelsmu committed:
update workflow

hamelsmu committed:
Merge branch 'master' of https://github.com/hamelsmu/fast_template

hamelsmu committed:
remove checkpoint

hamelsmu commented on .github/workflows/nb2post.yaml:
```diff
@@ -0,0 +1,41 @@
+        python-version: '3.6'
+        architecture: 'x64'
+    
+      # These are the minimal steps I could come up with install nbdev in the actions runner
```
@sgugger perhaps you would be interested in my install steps, I tried to make it as minimal as possible but these are the steps I found to make it work for me.  It still installs very very fast, though so no worries.

hamelsmu committed:
Update nb2post.yaml

Added magic commit message command

hamelsmu commented on .github/workflows/nb2post.yaml:
```diff
@@ -0,0 +1,42 @@
+
+jobs:
+  convert-notebook:
+    if: contains(join(github.event.commits.*.message), '/sync')
```
FYI this is one of the coolest parts about Actions which allow you to introspect the metadata of the events that trigger workflows.  In this line, we are checking to see if any of the commit messages in the push that triggers the workflow contain the string `/sync`

jph00 commented:
This is looking great @hamelsmu ! :) My only suggestion is to remove the sample notebook. Instead, put a README in the _notebooks folder telling people what that folder is there for, and how to use it.

Most folks using fast_template won't be blogging with notebooks, and won't know what they are, so I don't want to have them see a sample notebook post appear in their initial blog - I think that would be a bit intimidating!

hamelsmu committed:
remove pillow from setup

hamelsmu committed:
save intermediate README

hamelsmu commented:
Sounds good, still working on it will let you know when it's ready for review (I'll finish it by end of today)

hamelsmu committed:
update README

hamelsmu committed:
remove notebook post

hamelsmu committed:
correct small nits

hamelsmu committed:
lint, spell/grammar check

hamelsmu commented:
@jph00 I made the following changes you suggested:

- Added `_notebooks/README.md`
- Removed `_notebooks/*ipynb` and any correlated markdown files from `_posts/`
- Removed installation of PIL from setup steps

Please let me know if there is anything else I can add/change that would be useful!  Thank you 🙇

hamelsmu committed:
remove notebook

jph00 commented:
Fantastic!
